libbeat/publisher/pipeline/doc.go

```diff
@@ -0,0 +1,53 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+/*
+Package pipeline implements event publishing pipelines used in libbeat. These pipelines
+are used to publish events from a source (e.g. an input in Filebeat) all the way to an output
+(e.g. Elasticsearch), via processors and a queue. A pipeline encapsulates all these components
+and abstracts over some key internal components as well to make event publishing possible.
+
+The main component, of course, is a *pipeline* itself. When a new pipeline is created
+a *client* is returned. Users of the pipeline can use this client to publish
+events to the pipeline.
+
+When creating a new pipeline, several configuration options may be specified: e.g.
+the *processors* that the event must be passed through, the *queue* implementation (in-memory
+or spool-to-disk) to use, and the *group of outputs* to which the pipeline should
+ultimately send the processed events.
+
+For managing the group of outputs, the pipeline internally creates an *output controller*. The
+output controller creates three main entities:
+- a group of *output workers*. Each output worker manages one output and tries to publish events
+to it.
+- a *consumer* whose job is to dequeue events from the Beat's queue and enqueue them to an
+internal *work queue*. This work queue is shared amongst all output workers. Each worker dequeues
+batches of events from the work queue and tries to publish them to its underlying output.
+- a *retryer* whose job is to retry *cancelled* batches of events. If an output worker or its
+underlying output fails to publish a batch of events, that batch is cancelled. Cancelling a batch
+sends it to the retryer. When the retryer receives a batch of cancelled events, it enqueues them to
```

Comment:
One can use `batch.Cancelled` and `batch.Retry`. In both cases the batch is returned to the `retryer`, that will forward batches again to the work queue. Each batch has a TTL counter. If the TTL reaches 0 all events in the batch are dropped. Calling `batch.Retry` decreases the TTL counter, while `batch.Cancelled` does not touch the TTL counter.

Note: failed batches are not directly added to the workqueue, as this can lead to deadlocks.